python/ruff/__main__.py

```diff
@@ -35,22 +52,24 @@ def find_ruff_bin() -> str:
 
     # Search for pip-specific build environments.
     #
+    # Expect to find ruff in <prefix>/pip-build-env-<rand>/overlay/bin/ruff
+    # Expect to find a "normal" folder at <prefix>/pip-build-env-<rand>/normal
+    #
     # See: https://github.com/pypa/pip/blob/102d8187a1f5a4cd5de7a549fd8a9af34e89a54f/src/pip/_internal/build_env.py#L87
     paths = os.environ.get("PATH", "").split(os.pathsep)
     if len(paths) >= 2:
-        first, second = os.path.split(paths[0]), os.path.split(paths[1])
-        # Search for both an `overlay` and `normal` folder within a `pip-build-env-{random}` folder. (The final segment
-        # of the path is the `bin` directory.)
+        maybe_overlay = get_last_three_path_parts(paths[0])
+        maybe_normal = get_last_three_path_parts(paths[1])
         if (
-            len(first) >= 3
-            and len(second) >= 3
-            and first[-3].startswith("pip-build-env-")
-            and first[-2] == "overlay"
-            and second[-3].startswith("pip-build-env-")
-            and second[-2] == "normal"
+            len(maybe_normal) >= 3
+            and maybe_normal[-1].startswith("pip-build-env-")
```

Comment:
Isn't this wrong? The path is like `<prefix>/pip-build-env-<rand>/overlay/bin/ruff`. So then `maybe_normal` would be `['ruff', 'bin', 'overlay']`, right?

Reply:
The paths from `$PATH` don't include the actual executable. For instance for `maybe_overlay` the path we parse would look like `<prefix>/pip-build-env-<rand>/overlay/bin`, so `maybe_overlay` is `["bin", "overlay", "pip-build-env-<rand>"]`